video/out/drm_atomic.c

```diff
@@ -458,3 +497,113 @@ bool drm_mode_destroy_blob(int fd, struct drm_mode *mode)
+    p->data.metadata_type = DRM_STATIC_METADATA_TYPE1;
+    p->data.hdmi_metadata_type1.metadata_type = DRM_STATIC_METADATA_TYPE1;
+    struct mp_csp_primaries colors = mp_get_csp_primaries(color->primaries);
+
+    switch(color->gamma) {
+        case MP_CSP_TRC_BT_1886:
+            p->data.hdmi_metadata_type1.eotf = DRM_METADATA_EOTF_TRADITIONAL_HDR;
+            break; 
+        case MP_CSP_TRC_PQ:
+            p->data.hdmi_metadata_type1.eotf = DRM_METADATA_EOTF_SMPTE_2084;
+            break;
+        case MP_CSP_TRC_HLG:
+            p->data.hdmi_metadata_type1.eotf = DRM_METADATA_EOTF_HLG;
+            break;
+        default:
+            // Default is SDR
+            p->data.hdmi_metadata_type1.eotf = DRM_METADATA_EOTF_TRADITIONAL_SDR;
+    }
+
+    if(p->data.hdmi_metadata_type1.eotf) {
+        // Convert to nits (cd/m2):
+        float min_luminance = 0.0;
+        float max_luminance = color->sig_peak * MP_REF_WHITE;
+        float max_cll = max_luminance;
+        float max_fall = max_luminance;
+        
+        // Range 0x0000 to 0xC350 maps to 0.0 - 1.0000. Step of 0.00002. Same for White Point
+        p->data.hdmi_metadata_type1.display_primaries[0].x = clamp16_round(colors.red.x);
+        p->data.hdmi_metadata_type1.display_primaries[0].y = clamp16_round(colors.red.y);
+        p->data.hdmi_metadata_type1.display_primaries[1].x = clamp16_round(colors.green.x);
+        p->data.hdmi_metadata_type1.display_primaries[1].y = clamp16_round(colors.green.y);
+        p->data.hdmi_metadata_type1.display_primaries[2].x = clamp16_round(colors.blue.x);
+        p->data.hdmi_metadata_type1.display_primaries[2].y = clamp16_round(colors.blue.y);
+
+        // White point
+        p->data.hdmi_metadata_type1.white_point.x = clamp16_round(colors.white.x);
+        p->data.hdmi_metadata_type1.white_point.y = clamp16_round(colors.white.y);
+
+        // Min Luminance: 16-bit value in units of 0.0001 cd/m2, where 0x0001 represents 0.0001 cd/m2 and 0xFFFF represents 6.5535 cd/m2.
+        p->data.hdmi_metadata_type1.min_display_mastering_luminance = clamp16(floorf(min_luminance * 10000.0));
+
+        // Max Luminance: 16-bit value in units of 1 cd/m2, where 0x0001 represents 1 cd/m2 and 0xFFFF represents 65535 cd/m2.
+        p->data.hdmi_metadata_type1.max_display_mastering_luminance = clamp16(ceilf(max_luminance));
+
+        // Max Content Light Level: 16-bit value in units of 1 cd/m2, where 0x0001 represents 1 cd/m2 and 0xFFFF represents 65535 cd/m2.
+        p->data.hdmi_metadata_type1.max_cll = clamp16(ceilf(max_cll));
+
+        // Max Frame Average Light Level: 16-bit value in units of 1 cd/m2, where 0x0001 represents 1 cd/m2 and 0xFFFF represents 65535 cd/m2.
+        p->data.hdmi_metadata_type1.max_fall = clamp16(ceilf(max_fall));
+
+        drmModeCreatePropertyBlob(ctx->connector->fd, &p->data, sizeof p->data, &p->blob_id);
+        drmModeAtomicAddProperty(ctx->request, ctx->connector->id, property_hdr_id, p->blob_id);
+    }
```

Comment:
is it intentional that this function doesn't actually set anything if no EOTF is present?

Reply:
Yes, SDR is the fallback, so no metadata is sent.

Reply:
I figured out why this is broken: if you watch an HDR video and then SDR in the same mpv instance the `HDR_OUTPUT_METADATA` will never be reset